core/domain/html_cleaner.py

```diff
@@ -447,32 +517,81 @@ def validate_textangular_format(html_list, run_migration=False):
         soup = bs4.BeautifulSoup(
             soup_data.replace('<br>', '<br/>'), 'html.parser')
 
-        # Text with no parent tag is also invalid.
-        for content in soup.contents:
-            if not content.name:
-                err_dict['strings'].append(html_data)
-                break
-
-        for tag in soup.findAll():
-            # Checking for tags not allowed in RTE.
-            if tag.name not in allowed_tag_list:
-                if 'invalidTags' in err_dict:
-                    err_dict['invalidTags'].append(tag.name)
-                else:
-                    err_dict['invalidTags'] = [tag.name]
-                err_dict['strings'].append(html_data)
-            # Checking for parent-child relation that are not
-            # allowed in RTE.
-            parent = tag.parent.name
-            if (tag.name in allowed_tag_list) and (
-                    parent not in allowed_parent_list[tag.name]):
-                if tag.name in err_dict:
-                    err_dict[tag.name].append(parent)
-                else:
-                    err_dict[tag.name] = [parent]
+        is_invalid = validate_soup_for_textangular(soup, err_dict)
+
+        if is_invalid:
+            err_dict['strings'].append(html_data)
+
+        for collapsible in soup.findAll('oppia-noninteractive-collapsible'):
+            # content is taken from [-1:1] since quotes are unescaped
+            # and are treated as part of html content.
+            content_html = html_unescaper(
```

Comment:
See above; I'm a bit dubious about the different treatment of collapsible and tabs. For one thing, I don't think there's a corresponding "special case" in the frontend?

Reply:
Answered above!